src/docs/sphinx/dev_guide/tensor_dual.rst

```diff
@@ -0,0 +1,289 @@
+
+      tensor < double, 3 > u = {1.0, 2.0, 3.0};
+      tensor < double, 3 > v = {4.0, 5.0, 6.0};
+      tensor < double, 3 > sum = u + v;
+      tensor < double, 3 > weighted_sum = 3.0 * u - v / 7.0;
+      weighted_sum += sum;
+
+-  ``tensor`` supports many common mathematical functions:
+
+   .. code-block:: cpp
+
+      tensor < double, 3, 3 > I = Identity<3>();
+      tensor < double, 3, 3 > grad_u = {...};
+      tensor < double, 3, 3 > strain = 0.5 * (grad_u + transpose(grad_u)); // or, equivalently, sym(grad_u)
+      tensor < double, 3, 3 > stress = lambda * tr(strain) * I + 2.0 * mu * strain;
+
+      tensor < double, 3 > traction = dot(stress, normal);
+      tensor < double, 2, 3 > J = {...};
+      double dA = sqrt(det(dot(J, transpose(J))));
+      tensor < double, 3 > force = traction * dA;
+
+-  ``tensor`` supports useful shortcuts:
+
+   .. code-block:: cpp
+
+      // class template argument deduction: A is deduced to have type tensor<double,2,2>
+      tensor A = {{{1.0, 2.0}, {2.0, 3.0}}}; 
+
+      // create tensors from index notation expressions, B has type tensor<double,3,3,3,3>
+      tensor B = make_tensor<3,3,3,3>([](int i, int j, int k, int l){
+        return 0.5 * ((i == j) * (k == l) + (i == l) * (k == j));
+      });
+
+      // slicing: get and set specific subtensors
+      tensor< double, 3, 3 > C = B[0][2];
+      B[2][0][0] = C[1];
+
+      // access by operator() or operator[]
+      C[1][1] = 2.0;
+      C(2, 2) = 3.0;
+
+-   ``tensor`` arithmetic supports ``constexpr`` evaluation:
+
+   .. code-block:: cpp
+
+      constexpr tensor change_of_basis_matrix = {{
+        { 3.0, -2.0, 1.7},
+        { 2.0,  8.0, 1.7},
+        {-1.0,  4.0, 6.7}
+      }};
+
+      // express a quantity in a new basis
+      tensor v = dot(change_of_basis_matrix, u);
+
+      // modify the components in the new basis
+      v = f(v);
+
+      // precompute the inverse basis transformation at compile time
+      constexpr tensor inverse_change_of_basis_matrix = inv(change_of_basis_matrix);
+
+      // convert the modified values back to the original basis
+      u = dot(inverse_change_of_basis_matrix, v);
+
+-  ``tensor`` only allows operations between operands of appropriate
+   shapes
+
+   .. code-block:: cpp
+
+      tensor< double, 3, 2 > A{};
+      tensor< double, 3 > u{};
+      tensor< double, 2 > v{};
+
+      auto uA = dot(u, A); // works, returns tensor< double, 2 >
+      auto Av = dot(A, v); // works, returns tensor< double, 3 >
+      auto Au = dot(A, u); // compile error: incompatible dimensions for dot product
+      auto vA = dot(v, A); // compile error: incompatible dimensions for dot product
+
+      auto w = u + v; // compile error: can't add tensors of different shapes
+
+      A[0] = v; // works, assign a new value to the first row of A
+      A[1] = u; // compile error: can't assign a vector with 3 components to a vector of 2 components
+
+.. _header-n157:
+
+Dual Number Class
+=================
```

Comment:
```suggestion
Dual Number Class
-----------------
```

(and for the remaining subheadings in this file and the other RST file)

^ the subheadings make the table of contents easier to read, see e.g. https://serac.readthedocs.io/en/docs-essman-weak_form/sphinx/dev_guide/weak_form.html

Reply:
Thanks, this was bothering me too. I'm actually generating these .rst files from markdown via pandoc, so now I need to figure out what markdown formatting produces this appropriate subheading style.